pkg/storage/formattedstore/batchwrite.go

```diff
@@ -125,57 +130,58 @@ func (b *BatchService) Get(k string) ([]byte, error) {
 
 // Put in batch value.
 func (b *BatchService) Put(s batchStore, k string, v []byte) error {
-	b.RLock()
-	size := len(b.values)
-	b.RUnlock()
-
-	if size >= b.sizeLimit {
-		if err := b.flush(); err != nil {
-			return err
-		}
-	}
-
 	addBatch := &batch{keyID: k, value: v, isDeleted: false, addEncryptedIndicesFunc: s.AddEncryptedIndices}
+	start := time.Now()
 
-	b.Lock()
+	b.futureValuesLock.Lock()
 	b.futureValues[k] = addBatch
-	b.Unlock()
+	b.futureValuesLock.Unlock()
 
 	b.addBatchChan <- addBatch
 
+	logger.Infof("batch write put duration %s", time.Since(start))
+
 	return nil
 }
 
 // Delete value.
 func (b *BatchService) Delete(k string) {
 	addBatch := &batch{keyID: k, isDeleted: true}
 
-	b.Lock()
+	b.futureValuesLock.Lock()
 	b.futureValues[k] = addBatch
-	b.Unlock()
+	b.futureValuesLock.Unlock()
 
 	b.addBatchChan <- addBatch
 }
 
-func (b *BatchService) flush() error {
+// Flush data.
```

Comment:
How does the caller know when the data has _finished_ writing? If the application goes away (e.g., browser is closed - memory and go channels are gone), how does the caller recover when the application comes back?

Reply:
@troyronda the caller need to flush data when before finishing the flow